refactor of containers/stages/sparse

jdolence opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

This PR does several things that required touching a lot of files.  The major high-level changes include:

* Replaces Stages with ContainerCollection.  A MeshBlock now owns a ContainerCollection instead of a Container, which facilitates creating temporary buffers for multistage integration schemes, concurrent RHS evaluations, etc.
* Refactors SparseVariable to lift the map associating a string label with a sparse dataset up into the Container, and leaving the map associating integer sparse IDS with data in the SparseVariable class.
* Replaces AthenaArray with ParArrayND everywhere.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

Addresses #95, #88, #66, and #63.  Also relevant for #64.

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint. (Not all files, but the main targets.)
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.

pgrete commented:
Do you know where it went wrong originally?

pgrete commented:
Second question: What's your expected timeline to remove the WIP (or what's missing that you'd like to address before review)?
Especially, all the tedious cleanup work you did would be nice in master earlier than later.
I'm happy to review once this is good to go.

junghans commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 15
- Added 3
           

Clones added
============
- example/advection/advection.cpp  10
- example/advection/main.cpp  2
- example/face_fields/face_fields_example.cpp  2
- src/mesh/meshblock.cpp  2
- example/calculate_pi/pi.cpp  2
         

Clones removed
==============
+ src/outputs/athena_hdf5_C.cpp  -2
+ src/mesh/mesh.cpp  -2
+ src/interface/Container.cpp  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/lanl/parthenon/pullRequest?prid=5271978&bid=17267944)

junghans commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 8
           

Clones added
============
- example/face_fields/face_fields_example.cpp  2
- src/coordinates/coordinates.cpp  10
- example/advection/advection.cpp  10
- tst/unit/test_pararrays.cpp  10
- src/reconstruct/dc.cpp  1
- example/advection/main.cpp  2
- example/calculate_pi/pi.cpp  2
         

Clones removed
==============
+ src/interface/Container.cpp  -13
+ src/mesh/mesh.cpp  -6
+ src/outputs/athena_hdf5_C.cpp  -2
+ src/mesh/meshblock.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/lanl/parthenon/pullRequest?prid=5271978&bid=17267944)

junghans commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 8
           

Clones added
============
- example/face_fields/face_fields_example.cpp  2
- example/advection/advection.cpp  10
- src/coordinates/coordinates.cpp  10
- tst/unit/test_pararrays.cpp  10
- example/calculate_pi/pi.cpp  2
- src/reconstruct/dc.cpp  1
- example/advection/main.cpp  2
         

Clones removed
==============
+ src/interface/Container.cpp  -13
+ src/outputs/athena_hdf5_C.cpp  -2
+ src/mesh/mesh.cpp  -6
+ src/mesh/meshblock.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/lanl/parthenon/pullRequest?prid=5271978&bid=17267944)

jdolence commented:
@pgrete @Yurlungur @nmsriram This PR is getting close. I've still got some cleanup to do and need tests of, for example, SparseVariable.  At 83 modified files, I'd love to not sit on this for too long because basically every other PR is a potential conflict to resolve.  I don't expect this PR to ever get whipped into perfect shape -- it's meant to be something we can/will build on -- so I think we're aiming for good enough on this one.  Sorry for its ridiculous size.

jdolence commented:
> @jdolence I started to look at the changes from a conceptual point of view.
> 
> Originally I thought that that
> 
>     * each `MeshBlock` (usually) owns at least one `Package` and one `Container`.

Keep in mind that the Packages_t object owned by each MeshBlock isn't really the actual physics packages.  It's really an info object.  The real reason each MeshBlock owns one is because it can hold parameters that control how physics packages behave.  If you look at the flux function in the advection example, you'll see that the velocity is pulled out of this object, for example.

> 
>     * A `Container` holds (physical) `Variables` that may be shared/reusded across different packages

yes

> 
>     * Packages themselves could be completely independent (say one package takes are of a hydro update and another package does an operator split gravity solve)

yes

> 
>     * The `ParthenonManager` is then responsible to process all packages of each  `MeshBlock` in each cycle by means of a `Driver` that is part of a `Package`.
> 

ParthenonManager doesn't control the Driver.  Think of ParthenonManager as setup and take down only.  Sandwiched between setup and take down is the driver which is responsible for coordinating all of the real work of the simulation.

> 
> Are these assumptions correct at high level?
> 
> Then my question would be where/how stages fit in here (especially across multiple packages that will reuse variables) as the stages are currently handled within the multistage driver.
> 

"Stages" has been replaced by ContainerCollection.  The idea is you have a "base" container that you can think of as holding the persistent, evolved state in the simulation.  When you do rk2, for example, you need a temporary copy of the conserved variables.  So you can add a container to the collection that holds this for you.  You might also want something like delta_state -- you can make a container for that too.  The ContainerCollection allows you to add these containers and access them by name.

> Slightly related (in the sense that it's a conceptual question) I see that, for example, the reconstruction function is assigned to each `MeshBlock`.
> In my mind it'd make more sense to move the "reconstruction" within a package so that (theoretically I could update different variables of the MeshBlock with different reconstruction functions or, in practice, use a lower order reconstruction in a predictor stage and a higher order reconstruction in the corrector stage).
> 

Yes everything should just be a package.  The reconstruction object owned by the MeshBlock is just an Athena++ holdover.  Just haven't moved it out.  I figure we will if/when we write the performance portable version.  That should just be a package like any other.

> I guess where I'm conceptually going is move even more of the functionality that in the original Athena++ code was tightly coupled to a `MeshBlock` to a `Package` that is flexible in its configuration (pertains, e.g., also to the EOS or Riemann solver in our application).

Yes the idea is to decouple MeshBlock from what one might define as a package as much as possible.

Yurlungur commented:
> @pgrete @Yurlungur @nmsriram This PR is getting close. I've still got some cleanup to do and need tests of, for example, SparseVariable. At 83 modified files, I'd love to not sit on this for too long because basically every other PR is a potential conflict to resolve. I don't expect this PR to ever get whipped into perfect shape -- it's meant to be something we can/will build on -- so I think we're aiming for good enough on this one. Sorry for its ridiculous size.

I added us as reviewers. And I'll start poking around.

I think it would be wise to merge in my `ParArrayND` branch before we merge this in, since you're building off of it. And that will result in some merge conflicts, since you started using `ParArrayND` before it underwent a fair amount of review.

nmsriram commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  const auto& refine_tol = pkg->Param<Real>("refine_tol");
+  const auto& derefine_tol = pkg->Param<Real>("derefine_tol");
+
+  if (vmax > refine_tol && vmin < derefine_tol) return 1;
```
Should we be using named constants or an enum for AMR flags such as REFINE, DEREFINE, NONE instead of 1, -1, and 0 respectively?

jdolence replied:
Good idea.  Added enum class AmrTag : int {derefine=-1, same=0, refine=1}.  So now instead of return 1, it's return AmrTag::refine.  Better?

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+TaskStatus UpdateContainer(MeshBlock *pmb, int stage, std::vector<std::string>& stage_name, Integrator* integrator) {
+  //const Real beta = stage_wghts[stage-1].beta;
+  const Real beta = integrator->_beta[stage-1];
+  Container<Real>& base = pmb->real_containers.Get();
+  Container<Real>& cin = pmb->real_containers.Get(stage_name[stage-1]);
+  Container<Real>& cout = pmb->real_containers.Get(stage_name[stage]);
+  Container<Real>& dudt = pmb->real_containers.Get("dUdt");
+  parthenon::Update::AverageContainers(cin, base, beta);
+  parthenon::Update::UpdateContainer(cin, dudt, beta*pmb->pmy_mesh->dt, cout);
```
These lines probably deserve an explanatory comment. They're hard to read, especially for somebody who doesn't know what these methods do. Even a simple equation would be helpful. e.g.,
```C++
  // Performs one stage in multistage integration
  Container<Real>& base = pmb->real_containers.Get();
  Container<Real>& cin = pmb->real_containers.Get(stage_name[stage-1]);
  Container<Real>& cout = pmb->real_containers.Get(stage_name[stage]);
  Container<Real>& dudt = pmb->real_containers.Get("dUdt");
  // cin = beta*cin + (1 - beta)*base
  parthenon::Update::AverageContainers(cin, base, beta);
  // cout = cin + beta*dt*dudt
  parthenon::Update::UpdateContainer(cin, dudt, beta*pmb->pmy_mesh->dt, cout);
```

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+// function.                                       *//
+// *************************************************//
+// first some helper tasks
+TaskStatus UpdateContainer(MeshBlock *pmb, int stage, std::vector<std::string>& stage_name, Integrator* integrator) {
```
Since in this case, updating a container is integration, I wonder if this method could have a more descriptive name, like `Substep` or something.

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+
+TaskList AdvectionDriver::MakeTaskList(MeshBlock *pmb, int stage) {
+  TaskList tl;
+  // we're going to populate our last with multiple kinds of tasks
```
```suggestion
 // we're going to populate our list with multiple kinds of tasks
```

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  TaskList tl;
+  // we're going to populate our last with multiple kinds of tasks
+  // these lambdas just clean up the interface to adding tasks of the relevant kinds
+  auto AddMyTask = [&tl, pmb, stage, this] (BlockStageNamesIntegratorTaskFunc func, TaskID dep) {
```
Why not `AddIntegratorTask`?

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  if (stage == 1) {
+    Container<Real>& base = pmb->real_containers.Get();
+    pmb->real_containers.Add("dUdt", base);
+    for (int i=1; i<integrator->_nstages; i++) pmb->real_containers.Add(stage_name[i], base);
```
If `_nstages` is intended to be used this way, the prefix should change and it should be named `nstages`.

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+    for (int i=1; i<integrator->_nstages; i++) pmb->real_containers.Add(stage_name[i], base);
+  }
+
+  Container<Real>& sc0  = pmb->real_containers.Get(stage_name[stage-1]);
+  Container<Real>& sc1  = pmb->real_containers.Get(stage_name[stage]);
+  Container<Real>& dudt = pmb->real_containers.Get("dUdt");
```
This could use some explanation too. i.e., that `sc0` is the last subcycle and that `sc1` is the current one. Just to clarify how this works when we're doing multistage integration.

jdolence replied:
added some comments

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  Container<Real>& dudt = pmb->real_containers.Get("dUdt");
+
+
+  auto start_recv = AddContainerTask([](Container<Real>& rc) {
```
Lambdas for tasks are 👍 👍 👍

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  }, advect_flux, sc0);
+
+  auto recv_flux = AddContainerTask([](Container<Real>& rc) {
+      if (!rc.ReceiveFluxCorrection()) return TaskStatus::fail;
```
Some clarification on the meaning of `TaskStatus::fail` might be helpful. e.g., is it "this task will be done again until it succeeds."? Or is it "FATAL ERROR"?

jdolence replied:
I've renamed the ```TaskStatus``` entries.  It's now ```fail``` which indicates something about the task actually cannot be completed, ```complete``` meaning it's done, and ```incomplete``` meaning try again.  Is that clear enough?  Eventually, this should just be documented.

Yurlungur replied:
Yep that works for me.

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+    if ( !rc.ReceiveBoundaryBuffers() ) return TaskStatus::fail;
+    return TaskStatus::success;
+  }, send, sc1);
+  auto setC = AddContainerTask([](Container<Real>& rc) {
```
`setC` is a confusing name for this task

jdolence replied:
renamed.

Yurlungur commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+    }, fill_derived, pmb);
+
+  }
+  return tl;
```
It occurs to me that a lot of tasks in the task list are going to look identical for a lot of codes, which is why, of course, we provide methods in `parthenon` to perform them. But I wonder if there could be a method in `MultiStageIntegrator` to add all of them at once?

jdolence replied:
problem is dependencies.  how do these commonly performed tasks fit in with the rest of the stuff that needs to happen?

Yurlungur replied:
I feel like at least some of them depend on nothing, one unknown task, or only each other, which might be consistent across a wide range of methods. So there might be something that can be done to stitch them in piecemale. Whether or not that's worth doing, I don't know. It's probably a question for another time.

nmsriram replied:
@Yurlungur is it possible to provide pre-canned tasks in multistage integrator, similar to the methods in parthenon.  This also opens the door to behind-the-scenes optimization for common tasks.  I actually like the specification of the dependencies in the place where they are referenced and wouldn't like to muck that up.

Yurlungur replied:
Ooh, I like that @nmsriram. Maybe we could provide methods that let you do something like
```C++
auto recv = AddRecvGhostsTask(dependency);
```

jdolence replied:
@nmsriram @Yurlungur Not sure how to get exactly that API, but I have been meaning to make tasks out of a lot of the generic parthenon functionality.  I've pushed a step in this direction, providing tasks for Container mediated MPI ops.  I'm not sure the static member function approach I took is necessarily the best, but I don't hate it.  Thoughts?

Yurlungur replied:
I like that, @jdolence. I'll take a more careful look at the API as you implemented.

Yurlungur commented on example/advection/main.cpp:
```diff
@@ -0,0 +1,53 @@
+  using parthenon::ParthenonStatus;
+  ParthenonManager pman;
+
+  auto manager_status = pman.ParthenonInit(argc, argv);
+  if (manager_status == ParthenonStatus::complete) {
+    pman.ParthenonFinalize();
+    return 0;
+  }
+  if (manager_status == ParthenonStatus::error) {
+    pman.ParthenonFinalize();
+    return 1;
+  }
+
+  AdvectionDriver driver(pman.pinput.get(), pman.pmesh.get(), pman.pouts.get());
+
+  // start a timer
+  pman.PreDriver();
+
+  auto driver_status = driver.Execute();
+
+  // Make final outputs, print diagnostics
+  pman.PostDriver(driver_status);
+
+  // call MPI_Finalize if necessary
+  pman.ParthenonFinalize();
```
So no `Kokkos::Views` can be created before `ParthenonInit` and they must all be gone by `ParthenonFinalize`. I realize that the `init` and `finalize` methods ensure this happens, but I'd still feel more comfortable if they wrapped scope. Especially for an example.

jdolence replied:
I'm not sure I understand what you have in mind here.  Do you want { } somewhere?  We could put them between PathenonInit and ParthenonFinalize, but that's a lie.  ParthenonInit, in particular, will certainly have new ```Kokkos::Views``` created.

Yurlungur replied:
Yeah i wanted them between Parthenon init and Parthenon finalize.

Maybe we don’t need that. Instead perhaps a comment explaining where kokkos init and finalize are and what that implies about views would suffice.

Yurlungur commented on example/advection/parthinput.advection:
```diff
@@ -0,0 +1,64 @@
+<output1>
+file_type = hdf5
+dt = 0.05
+variable = prim
```
We should really get rid of this. It doesn't do anything, since which variables are output are set by `Metadata`.

jdolence replied:
That's true, but this comment actually got me thinking.  It would be nice at some point to have some control over which variables had Metadata::Graphics set from the input file.  Not sure what the right model is here.  Just a list of variables, perhaps?

Yurlungur replied:
I think a list of variables would suffice. We could implement this pretty easily I think... just add the graphics metadata flag to variables tagged for output.

forrestglines replied:
> We should really get rid of this. It doesn't do anything, since which variables are output are set by `Metadata`.

Can you still control which variables are written to disk through the parameter file? That seems like a very necessary feature, we don't always want the same variables written to disk for every simulation. That shouldn't be hard coded.

jdolence replied:
Yes clearly we need control of this.  It's just a matter of setting the Metadata::Graphics flag for each variable you want to show up.  Right now this is hard coded in the examples, but only because we haven't written the little bit of code that parses some field in the input file and sets the flag.

jdolence replied:
Made an issue: #98

Yurlungur commented on example/calculate_pi/pi.cpp:
```diff
@@ -27,14 +27,14 @@ namespace parthenon {
   // only have one package for this app, but will typically have more things added to
   packages["PiCalculator"] = PiCalculator::Initialize(pin.get());
-  return std::move(packages);
+  return packages;
```
👍

Yurlungur commented on example/face_fields/face_fields_example.cpp:
```diff
@@ -148,13 +149,13 @@ parthenon::TaskStatus FaceFields::fill_faces(parthenon::MeshBlock* pmb) {
   auto& face = rc.GetFace("f.f.face_averaged_value");
   // fill faces
-  for (int e = 0; e < face.Get(1).GetDim4(); e++) {
+  for (int e = 0; e < face.Get(1).GetDim(4); e++) {
```
👍

Yurlungur commented on src/CMakeLists.txt:
```diff
@@ -66,6 +68,7 @@ set(NFIELD_VARIABLES 0) # TODO: Remove
 set(NWAVE_VALUE 5) # TODO: Remove
 set(NUMBER_GHOST_CELLS 2) # TODO: Make this a CMake option and check it
 
+
```
```suggestion
```

Yurlungur commented on src/CMakeLists.txt:
```diff
@@ -41,6 +41,8 @@ if (${Kokkos_ENABLE_CUDA})
   set(PAR_LOOP_LAYOUT_VALUES "MANUAL1D_LOOP;MDRANGE_LOOP;TPTTR_LOOP;TPTTRTVR_LOOP"
     CACHE STRING "Possible loop layout options.")
 
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --expt-relaxed-constexpr")
+
```
Uh oh. I think I added this twice somehow. This one shouldn't be here.
```suggestion
```

Yurlungur commented on src/athena.hpp:
```diff
@@ -160,28 +157,28 @@ enum class UserHistoryOperation {sum, max, min};
 
 using BValFunc = void (*)(
-    MeshBlock *pmb, Coordinates *pco, AthenaArray<Real> &prim, FaceField &b,
+    MeshBlock *pmb, Coordinates *pco, ParArrayND<Real> &prim, FaceField &b,
```
It doesn't hurt to pass `ParArrayND` by reference. But you don't need to, since they shallow copy.

Yurlungur commented on src/bvals/bvals.hpp:
```diff
@@ -108,9 +109,9 @@ class BoundaryValues : public BoundaryBase, //public BoundaryPhysics,
+  std::vector<std::shared_ptr<BoundaryVariable>> bvars;
   // subset of bvars that are exchanged in the main TimeIntegratorTaskList
-  std::vector<BoundaryVariable *> bvars_main_int;
+  std::vector<std::shared_ptr<BoundaryVariable>> bvars_main_int;
```
👍

Yurlungur commented on src/bvals/bvals_base.cpp:
```diff
@@ -100,8 +100,8 @@ BoundaryBase::BoundaryBase(Mesh *pm, LogicalLocation iloc, RegionSize isize,
     int nc1 = block_size_.nx1 + 2*NGHOST;
-    sarea_[0].NewAthenaArray(nc1);
-    sarea_[1].NewAthenaArray(nc1);
+    sarea_[0].NewParArrayND(nc1);
+    sarea_[1].NewParArrayND(nc1);
```
We should move towards naming these guys when we do this. Obviously that's something that will take a while to fix.

jdolence replied:
replaced all of these calls everywhere.

Yurlungur commented on src/outputs/athena_hdf5_C.cpp:
```diff
@@ -72,15 +72,15 @@ static void writeXdmfArrayRef(std::ofstream& fid, const std::string& prefix,
 // XDMF subroutine to write a variable that reads from a HDF file
-static void writeXdmfVariableRef(std::ofstream& fid, const std::string& prefix,
+/*static void writeXdmfVariableRef(std::ofstream& fid, const std::string& prefix,
 				 const std::string& hdfPath, const std::string& label,
 				 const hsize_t* dims, const int& ndims,
 				 const std::string& theType, const int& precision) {
   std::string mystr = prefix + "<Attribute Name=\"" + label + R"(" Center="Cell">)"+ '\n';
   mystr += stringXdmfArrayRef(prefix+"  ", hdfPath, label, dims, ndims, theType, precision);
   mystr += prefix + "</Attribute>\n";
   fid << mystr << std::flush;
-}
+}*/
```
Let's just kill it
```suggestion
```

Yurlungur commented on src/outputs/athena_hdf5_C.cpp:
```diff
@@ -294,7 +294,7 @@ void ATHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, bool flag) {
   MeshBlock *pmb = pm->pblock;
   int max_blocks_global = pm->nbtotal;
-  int max_blocks_local = pm->nblist[Globals::my_rank];
+  //int max_blocks_local = pm->nblist[Globals::my_rank];
```
```suggestion
```

Yurlungur commented on src/mesh/meshblock_tree.hpp:
```diff
@@ -26,7 +26,7 @@
 // Athena++ headers
 #include "athena.hpp"
-#include "athena_arrays.hpp"
+#include "parthenon_arrays.hpp"
```
just remove this include?

Yurlungur commented on src/mesh/meshblock.cpp:
```diff
@@ -277,6 +275,7 @@ MeshBlock::MeshBlock(int igid, int ilid, Mesh *pm, ParameterInput *pin,
 
   return;
 }
+#endif
```
If this whole block of code is commented out, can we delete it?

pgrete replied:
Eventually, we need it back in in order to allow for restarting simulations.

Yurlungur commented on src/mesh/meshblock.cpp:
```diff
@@ -286,18 +285,19 @@ MeshBlock::~MeshBlock() {
   if (nreal_user_meshblock_data_ > 0) delete [] ruser_meshblock_data;
   if (nint_user_meshblock_data_ > 0) delete [] iuser_meshblock_data;
-
+#endif
```
delete this block of code?

jdolence replied:
done.

Yurlungur commented on src/mesh/meshblock.cpp:
```diff
@@ -307,13 +307,13 @@ void MeshBlock::AllocateRealUserMeshBlockDataField(int n) {
   }
   nreal_user_meshblock_data_ = n;
-  ruser_meshblock_data = new AthenaArray<Real>[n];
+  ruser_meshblock_data = nullptr;//new ParArrayND<Real>[n];
```
Does the `user_meshblock_data` mechanism still work? And do we want it to? At some level, containers and variables do this job better.

pgrete replied:
I also think that the `Variables` and `Containers` provide the abstraction that was originally meant to be satisfied by the user meshblock data.
We could think about adding a "user" variable within the problem generator in one of the examples.

jdolence replied:
removed.

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
 #pragma omp for
       for (int i=0; i<nmb; ++i) {
-        pmb_array[i]->real_container.ReceiveAndSetBoundariesWithWait();
+      //std::cout << call << std::endl;
+        //pmb_array[i]->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
       for (int i=0; i<nmb; ++i) {
-        pmb_array[i]->real_container.SetBoundaries();
-        pmb_array[i]->real_container.ClearBoundary(BoundaryCommSubset::mesh_init);
+     // std::cout << call << std::endl;
+       // pmb_array[i]->real_containers.Get().print();
+        //pmb_array[i]->real_containers.Get().SetBoundaries();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
+       // pmb_array[i]->real_containers.Get().print();
+        //pmb_array[i]->real_containers.Get().SetBoundaries();
+        pmb_array[i]->real_containers.Get().ClearBoundary(BoundaryCommSubset::mesh_init);
+      //std::cout << call+0.5 << std::endl;
+        //pmb_array[i]->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
 #pragma omp for
       for (int i=0; i<nmb; ++i) {
         auto &pmb = pmb_array[i];
+     // std::cout << call << std::endl;
+       // pmb->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
 #pragma omp for
       for (int i=0; i<nmb; ++i) {
-        pmb_array[i]->real_container.SendBoundaryBuffers();
+      //std::cout << call << std::endl;
+        //pmb_array[i]->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
 #pragma omp for
       for (int i=0; i<nmb; ++i) {
-        pmb_array[i]->real_container.StartReceiving(BoundaryCommSubset::mesh_init);
+      //std::cout << call << std::endl;
+        //pmb_array[i]->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
       pmb->pbval->SetupPersistentMPI();
-      pmb->real_container.SetupPersistentMPI();
+      pmb->real_containers.Get().SetupPersistentMPI();
+      //std::cout << call+0.5 << std::endl;
+      //pmb->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1250,47 +1255,64 @@ void Mesh::Initialize(int res_flag, ParameterInput *pin) {
     for (int i=0; i<nmb; ++i) {
       MeshBlock *pmb = pmb_array[i];
       // BoundaryVariable objects evolved in main TimeIntegratorTaskList:
+      //std::cout << call << std::endl;
+      //pmb->real_containers.Get().print();
```
```suggestion
```

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -824,6 +826,7 @@ Mesh::Mesh(ParameterInput *pin, IOWrapper& resfile,
   delete [] offset;
 
 }
+#endif
```
Can we delete this block of code?

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+
+# SparseVariable
+
+The ```SparseVariable``` class is designed to support multi-component state where not all components may be present and therefore need to be stored.  At its core, the data is represented using a map that associates an integer ID to a ```std::shared_ptr<CellVariable<T>>```.  Since all ```CellVariable``` entries are assumed to have identical ```Metadata``` flags, the class provides an ```isSet``` m ...
````
Is a GetData method really necessary when the ParArrayND object is public?

Yurlungur replied:
It’s not necessary but i think it would be convenient.

jdolence commented on src/bvals/cc/bvals_cc.cpp:
```diff
@@ -46,18 +46,18 @@ namespace parthenon {
-    MeshBlock *pmb, AthenaArray<Real> *var, AthenaArray<Real> *coarse_var,
-    AthenaArray<Real> *var_flux)
+    MeshBlock *pmb, ParArrayND<Real> var, ParArrayND<Real> coarse_var,
+    ParArrayND<Real> *var_flux)
```
I almost made that change myself actually...

jonahm-LANL commented:
It’s not necessary, but it would provide some syntactic sugar, which i think would have value.


Sent from my phone. Please excuse typos.



From: jdolence <notifications@github.com<mailto:notifications@github.com>>
Date: Saturday, Apr 04, 2020, 9:12 PM
To: lanl/parthenon <parthenon@noreply.github.com<mailto:parthenon@noreply.github.com>>
Cc: Subscribed <subscribed@noreply.github.com<mailto:subscribed@noreply.github.com>>
Subject: [EXTERNAL] Re: [lanl/parthenon] WIP: refactor of containers/stages/sparse (#75)


@jdolence commented on this pull request.

________________________________

In docs/interface/state.md<https://github.com/lanl/parthenon/pull/75#discussion_r403642273>:

> +| ```std::string _label``` | The name of the variable. |
+
+The class provides a nontrivial copy constructor.  Given a ```CellVariable```, the copy constructor allocates new storage for a ```CellVariable``` with the same name and ```Metadata``` but does not copy the values into this new storage.  The main motivation for this is to support methods like multi-stage integration that require multiple registers for independent state.
+
+Additionally, the class overloads the ```()``` operator to provide convenient access to the ```data``` array, though this may be less efficient than operating directly on ```data``` or a reference/copy of that array.
+
+Finally, the ```bool isSet(const MetadataFlag bit)``` member function provides a convenient mechanism to query whether a particular ```Metadata``` flag is set for the ```CellVariable```.
+
+
+# FaceVariable (Work in progress...)
+
+# EdgeVariable (Work in progress...)
+
+# SparseVariable
+
+The ```SparseVariable``` class is designed to support multi-component state where not all components may be present and therefore need to be stored.  At its core, the data is represented using a map that associates an integer ID to a ```std::shared_ptr<CellVariable<T>>```.  Since all ```CellVariable``` entries are assumed to have identical ```Metadata``` flags, the class provides an ```isSet``` member function identical to the ```CellVariable``` class that applies to all variables stored in the map.  The ```Get``` method takes an integer ID as input and returns a reference to the associated ```CellVariable```, or throws a ```std::invalid_argument``` error if it does not exist.  The ```GetVector``` method returns a dense ```std::vector```, eliminating the sparsity but also the association to particular IDs.  The ```GetIndex``` method provides the index in this vector associated with a given sparse ID, and returns -1 if the ID does not exist.


Is a GetData method really necessary when the ParArrayND object is public?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/lanl/parthenon/pull/75#discussion_r403642273>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOO5OJ2ATN4MUP2HURV53CLRK7ZKRANCNFSM4LT2KEPQ>.

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+| ```ParArrayND<T> data``` | Storage for the cell-centered associated with the object. |
+| ```ParArrayND<T> flux[3]``` | Storage for the face-centered intercell fluxes in each direction.<br>Only allocated for fields registered with the ```Metadata::Independent``` flag. |
+| ```ParArrayND<T> coarse_s``` | Storage for coarse buffers need for multilevel setups. |
+| ```std::array<int,6> _dims``` | Dimensions of the data. |
````
nope.  removed.

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+| ```ParArrayND<T> coarse_s``` | Storage for coarse buffers need for multilevel setups. |
+| ```std::array<int,6> _dims``` | Dimensions of the data. |
+| ```Metadata _m``` | See [here](Metadata.md). |
+| ```std::string _label``` | The name of the variable. |
````
nope.  removed.

forrestglines commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  CellVariable<Real>& v = rc.Get("advected");
+  Real vmin = 1.0;
+  Real vmax = 0.0;
+  for (int k=0; k<pmb->ncells3; k++) {
+    for (int j=0; j<pmb->ncells2; j++) {
+      for (int i=0; i<pmb->ncells1; i++) {
+        vmin = (v(k,j,i) < vmin ? v(k,j,i) : vmin);
+        vmax = (v(k,j,i) > vmax ? v(k,j,i) : vmax);
+      }
+    }
+  }
```
Since we only had one use of a reduction in the whole code, we used ``Kokkos::parallel_reduce`` instead of making an `athena_reduce`. We could potentially create a `par_reduce`, but until we know all the things we might need out of a reduction abstraction, we should just use `Kokkos::parallel_reduce`. 

For this loop we would probably use the `Kokkos:MinMax` reducer documented at https://github.com/kokkos/kokkos/wiki/Kokkos%3A%3AMinMax so that we make just one kernel call instead of two.

forrestglines commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,399 @@
+  ParArrayND<Real> dx1("dx1",pmb->ncells1);
+
+  // this is obviously overkill for this constant velocity problem
+  for (int k=ks; k<=ke; k++) {
+    for (int j=js; j<=je; j++) {
+      pmb->pcoord->CenterWidth1(k, j, is, ie, dx0);
+      pmb->pcoord->CenterWidth2(k, j, is, ie, dx1);
+      for (int i=is; i<=ie; i++) {
+        min_dt = std::min(min_dt, dx0(i)/std::abs(vx));
+        min_dt = std::min(min_dt, dx1(i)/std::abs(vy));
+      }
+    }
+  }
```
> We also will want `pmb->pcoord` to be decorated appropriately.

By "decorated" do you mean aliasing or getting the right pointer to `CenterWidth1` on the GPU?

Yurlungur replied:
I meant the latter.

forrestglines commented on src/interface/Update.cpp:
```diff
@@ -114,10 +108,10 @@ void UpdateContainer(Container<Real> &in, Container<Real> &dudt_cont,
   int nvars = cout_iter.vars.size();
 
   for (int n = 0; n < nvars; n++) {
```
This kernel is potentially going to hinder performance on the GPUs. We need to be able to run this in a single kernel like in K-Athena, putting `par_for` on the outermost loop, instead of having one kernel for each variable. We would also not want to have `Variable` objects within the kernel, we want to access clean `Kokkos:View`s or `ParArrayND`. Looping over an array of `Variable`s or `ParArrayND`s is probably also not going to be viable for performance.

Like in K-Athena, we would like all the variables in a mesh block to be in one contiguous data object that can be looped over.

jdolence replied:
I've been meaning to add an issue for a redesigned ContainerIterator that is GPU friendly.  I'll add it.  This is certainly beyond the scope of the PR, though.

nmsriram commented on example/advection/advection.hpp:
```diff
@@ -0,0 +1,94 @@
+
+#include <memory>
+
+#include "driver/driver.hpp"
+#include "driver/multistage.hpp"
+#include "interface/Container.hpp"
+#include "interface/StateDescriptor.hpp"
+#include "mesh/mesh.hpp"
+#include "task_list/tasks.hpp"
+
+using parthenon::MultiStageBlockTaskDriver;
+using parthenon::MeshBlock;
+using parthenon::Mesh;
+using parthenon::Outputs;
+using parthenon::BaseTask;
+using parthenon::Real;
+using parthenon::CellVariable;
+using parthenon::StateDescriptor;
+using parthenon::TaskStatus;
+using parthenon::TaskList;
+using parthenon::TaskListStatus;
+using parthenon::TaskID;
+using parthenon::Metadata;
+using parthenon::Params;
+using parthenon::Container;
+using parthenon::ParameterInput;
+using parthenon::ParArrayND;
+using parthenon::Integrator;
+using parthenon::BlockStageNamesIntegratorTaskFunc;
+using parthenon::BlockStageNamesIntegratorTask;
+using parthenon::BlockTaskFunc;
+using parthenon::BlockTask;
+using parthenon::BoundaryValues;
+using parthenon::AmrTag;
```
At some point we should consider providing an include file that has a standard set of includes in something like packageInclude.hpp so that this code doesn't have to be replicated for every package.

Yurlungur replied:
I'm not sure I agree. I don't like generic includes, as they hide information. I think this boiler plate has some value as a comment.

AndrewGaspar replied:
The way Rust solves this is they have a "prelude" module for common names. So we could perhaps have something like:

```c++
namespace parthenon { namespace prelude {
// Re-exports
using Real = ::parthenon::Real;
using MeshBlock = ::parthenon::MeshBlock;
using ParArrayND = ::parthenon::PayArrayND;
} }
```

And then in the user code, you could write:
```c++
using namespace parthenon::prelude;
```

You should be careful, though, to keep that `prelude` namespace fairly lean - only things that are used by essentially ever single user. You could potentially split the preludes into something like `parthenon::task::prelude` and `parthenon::app::prelude` to distinguish being names commonly used by apps and names commonly used by task implementations.

nmsriram replied:
> I'm not sure I agree. I don't like generic includes, as they hide information. I think this boiler plate has some value as a comment.

Then why not just use everything from namespace parthenon? When the list of entities used from the namespace exceeds five or six, it starts to get unwieldy.  This would be no different than including a header file. My fear is that at some point folks will just copy and paste the entire 'using' list as their starting point and not prune it down to the ones used.  

I like the idea of splitting parthenon into sub-spaces and using those.  Some subsets that seem useful are Mesh, AMR, Task, and Types.

jdolence replied:
Made an issue: #100.

Yurlungur commented on src/interface/StateDescriptor.hpp:
```diff
@@ -55,23 +55,32 @@ class StateDescriptor {
+          m.Associate("");
+        }
       }
+      return true;
```
The `StateDescriptor` could probably use some documentation explaining its purpose... e.g., that it is primarily used to tell the mesh about the requirements of an individual package, e.g., it tells the meshblocks how to initialize storage.

jdolence replied:
done.

Yurlungur commented on src/interface/Variable.hpp:
```diff
@@ -164,58 +106,70 @@ class Variable : public AthenaArray<T> {
+  FaceArray<T> data;
+
  private:
+  std::array<int,6> _dims;
```
I'm torn about storing `_dims`, as this information is contained within `ParArrayND data`. But there is arithmetic involved in extracting it, where here there's none.

jdolence replied:
I removed it, but I'm second guessing myself.

forrestglines commented on tst/unit/test_container_iterator.cpp:
```diff
@@ -0,0 +1,139 @@
+    WHEN("We initialize all variables to zero") {
+      // set them all to zero
+      const CellVariableVector<Real>& cv = rc.GetCellVariableVector();
+      for (int n=0; n<cv.size(); n++) {
```
Thanks @jdolence ! This looks like a good starting point. The goal would be to move this layer of the looping into a kernel. Later on I'll add some performance tests  with some of the kernels that are likely to cause problems. Not in this PR though.

jdolence commented on example/advection/advection.hpp:
```diff
@@ -0,0 +1,92 @@
+#include "mesh/mesh.hpp"
+#include "task_list/tasks.hpp"
+
+using parthenon::MultiStageBlockTaskDriver;
```
Just for the record, I understand why we put everything in the parthenon namespace, but I hate all these lines, whether they're in a cpp or an hpp.  All it makes me want to do is
```c++
using namespace parthenon;
```
and move on with my life.  Perhaps the ```prelude``` approach discussed in another thread is a way out?

jdolence replied:
What is the appropriate way to write function declarations (for example) in a header with this restriction?  Having ```parthenon::``` all over the place is ugly and makes things much less readable.

jdolence replied:
Just pushed a possibly OK fix.  Please let me know.

pgrete commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -59,30 +59,30 @@ MeshRefinement::MeshRefinement(MeshBlock *pmb, ParameterInput *pin) :
   // KGF: probably don't need to preallocate space for pointers in these vectors
-  pvars_cc_.reserve(3);
-  pvars_fc_.reserve(3);
+  //pvars_cc_.reserve(3);
```
```suggestion
```

pgrete commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -59,30 +59,30 @@ MeshRefinement::MeshRefinement(MeshBlock *pmb, ParameterInput *pin) :
-  pvars_cc_.reserve(3);
-  pvars_fc_.reserve(3);
+  //pvars_cc_.reserve(3);
+  //pvars_fc_.reserve(3);
```
```suggestion
```

pgrete commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -59,30 +59,30 @@ MeshRefinement::MeshRefinement(MeshBlock *pmb, ParameterInput *pin) :
+  sarea_x3_[2][0].NewParArrayND(nc1);
+  sarea_x3_[2][1].NewParArrayND(nc1);
 
   // KGF: probably don't need to preallocate space for pointers in these vectors
```
```suggestion
```

jdolence commented on CMakeLists.txt:
```diff
@@ -122,8 +122,11 @@ if (${ENABLE_UNIT_TESTS} OR ${ENABLE_INTEGRATION_TESTS} OR ${ENABLE_REGRESSION_T
   add_subdirectory(tst)
 endif()
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
```
don't see this line.  I think your checkout was out of date.

jdolence commented on src/interface/SparseVariable.hpp:
```diff
@@ -31,9 +31,8 @@
-class SparseMap : public std::map<int, std::shared_ptr<Variable<T>>> {
- public:
-  Variable<T>& operator()(int m) {
+struct SparseMap : public std::map<int, std::shared_ptr<CellVariable<T>>> {
```
yes this is awful.  fixed.  thanks for reminding me about this nastiness.

jdolence commented on src/interface/SparseVariable.hpp:
```diff
@@ -54,126 +53,89 @@ class SparseMap : public std::map<int, std::shared_ptr<Variable<T>>> {
-                                   label +
-                                   " in SparseMap<T> container::Get() ");
+  CellVariable<T>& Get(const int index) {
+    if (_varMap.find(index) == _varMap.end()) {
+      throw std::invalid_argument("index " + std::to_string(index) +
+                                  "does not exist in SparseVariable");
     }
-    return _cellVars[label];
+    return *(_varMap[index]);
```
good catch.

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+| ```Metadata _m``` | See [here](Metadata.md). |
+| ```std::string _label``` | The name of the variable. |
+
+The class provides a nontrivial copy constructor.  Given a ```CellVariable```, the copy constructor allocates new storage for a ```CellVariable``` with the same name and ```Metadata``` but does not copy the values into this new storage.  The main motivation for this is to support methods like multi-stage integration that require multiple registers for independent state.
````
Done.

jdolence commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,417 @@
+
+  field_name = "one_minus_advected";
+  m = Metadata(
+    {Metadata::Cell, Metadata::Graphics, Metadata::Derived, Metadata::OneCopy}
```
Can we deal with this in #98?

jdolence commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,417 @@
+}
+
+
+TaskList AdvectionDriver::MakeTaskList(MeshBlock *pmb, int stage) {
```
added some comments

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+
+The ```ContainerCollection``` class is the highest level abstraction in parthenon's state management.  Each ```MeshBlock``` in a simulation owns a ```ContainerCollection``` that through the classes just described, manages all of the simulation data.  Every ```ContainerCollection``` is initialized with a ```Container``` named ```"base"```.  The ```Get``` function, when invoked without arguments, r ...
+
+The ```Add(const std::string& label, Container<T>& src)``` member function creates a new ```Container``` with the provided label.  This new ```Container``` is populated with all of the variables in ```src```.  When a variable has the ```Metadata::OneCopy``` flag set, the variables in the new ```Container``` are are just shallow copies from ```src```, i.e. no new storage for data is allocated, the ...
````
done.  Also added better error checking in the Add function.

jdolence commented on example/advection/advection.cpp:
```diff
@@ -0,0 +1,417 @@
+  auto update_container = AddMyTask(UpdateContainer, flux_div);
+
+  // update ghost cells
+  auto send = AddContainerTask(Container<Real>::SendBoundaryBuffersTask,
```
The problem with hiding them in the infrastructure is that you, as an application developer, need control over when things like communication are done.  Open to more discussion on the topic, but I think it's tough to hide this stuff effectively.

jdolence commented on docs/README.md:
```diff
@@ -51,9 +51,9 @@ A description of how to enable and extend the AMR capabilities of Parthenon is p
 
-Parthenon supports anonymous field variables. These variables can be registered
+Parthenon provides a convenient means of managing simulation data. Variables can be registered
 with Parthenon to have the framework automatically manage the field, including
```
Added some documentation on StateDescriptor, which I think addresses this.

jdolence commented on src/driver/multistage.hpp:
```diff
@@ -25,9 +25,9 @@ namespace parthenon {
 
 struct Integrator {
```
Added new documentation for drivers, which references Integrator.  Sufficient for now?

jdolence commented on src/interface/Container.cpp:
```diff
@@ -496,103 +390,65 @@ void Container<T>::ClearBoundary(BoundaryCommSubset phase) {
-  const int d4 = V.GetDim4();
+  const int d6 = V.GetDim(6);
+  const int d5 = V.GetDim(5);
+  const int d4 = V.GetDim(5);
```
This was definitely a bug, but just pointed out that we don't use this function so I removed it.

jdolence commented on src/interface/Variable.cpp:
```diff
@@ -168,10 +121,12 @@ void Variable<T>::allocateComms(MeshBlock *pmb) {
   //pmb->RegisterMeshBlockData(*this);
 
-  mpiStatus = true;
+  mpiStatus = false;
```
nope, fixed.  thanks -- good catch.

jdolence commented on src/interface/Container.cpp:
```diff
@@ -61,28 +63,34 @@ void Container<T>::Add(const std::string label,
                        const Metadata &metadata,
                        const std::vector<int> dims) {
   std::array<int, 6> arrDims;
```
The MeshBlock constructor calls Container<T>::Add which calls the constructor for CellVariable<T> (for example) in which a new ParArrayND object is created (and therefore memory allocated).

jdolence commented:
@forrestglines thanks for the comments.  Part of the hope with the abstractions is to provide a means of optimizing things under the hood.  The things you point out are reasonable examples of things like this that could (maybe should) be pursued.  I agree, though, these are for a later effort, not this PR.

jdolence commented on docs/interface/state.md:
````diff
@@ -0,0 +1,51 @@
+
+The class provides a nontrivial copy constructor.  Given a ```CellVariable```, the copy constructor allocates new storage for a ```CellVariable``` with the same name and ```Metadata``` but does not copy the values into this new storage.  The main motivation for this is to support methods like multi-stage integration that require multiple registers for independent state.
+
+Additionally, the class overloads the ```()``` operator to provide convenient access to the ```data``` array, though this may be less efficient than operating directly on ```data``` or a reference/copy of that array.
````
I don't think we intend CellVariable to ever be passed directly into a kernel, so this is likely not an issue.

jdolence commented on src/driver/driver.hpp:
```diff
@@ -54,20 +54,19 @@ class EvolutionDriver : public Driver {
     while (pmb != nullptr) {
-      task_lists[i] = driver->MakeTaskList(pmb, std::forward<Args>(args)...);
-      i++;
+      task_lists.push_back (driver->MakeTaskList(pmb, std::forward<Args>(args)...) );
```
Yeah I can't imagine this could matter, but at the very least let's only change if we actually see that it's a problem.

jdolence commented on src/driver/driver.hpp:
```diff
@@ -54,20 +54,19 @@ class EvolutionDriver : public Driver {
     int complete_cnt = 0;
     while (complete_cnt != nmb) {
-#pragma omp parallel for reduction(+ : complete_cnt) num_threads(nthreads) schedule(dynamic,1)
+//#pragma omp parallel for reduction(+ : complete_cnt) num_threads(nthreads) schedule(dynamic,1)
```
done

jdolence commented on src/interface/Metadata.hpp:
```diff
@@ -237,7 +237,12 @@ class Metadata {
+      return false;
+    }
+    return bits_[bit.flag_];
+  }
```
it did and it's merged now.

jdolence commented on src/interface/container_collection.cpp:
```diff
@@ -0,0 +1,61 @@
+template <typename T>
+void ContainerCollection<T>::Add(const std::string& name, Container<T>& src) {
+  // error check for duplicate names
+  if (containers_.find(name) != containers_.end()) {
```
this part was rewritten for other reasons and I think makes sense as is.

jdolence commented on src/interface/container_collection.hpp:
```diff
@@ -0,0 +1,68 @@
+  Container<T>& Get(const std::string& label) {
+    auto it = containers_.find(label);
+    if (it == containers_.end()) {
+      throw std::runtime_error("Container " + label + " does not exist in collection.");
```
what's wrong with throwing?

jdolence commented on src/bvals/cc/bvals_cc.cpp:
```diff
@@ -46,18 +47,18 @@ namespace parthenon {
 // constructor
 
 CellCenteredBoundaryVariable::CellCenteredBoundaryVariable(
```
Whatever approach we work out for ContainerIterator is likely to be adaptable to this problem as well.

jdolence commented on src/driver/multistage.cpp:
```diff
@@ -19,8 +19,8 @@ namespace parthenon {
 MultiStageDriver::MultiStageDriver(ParameterInput *pin, Mesh *pm, Outputs *pout) : EvolutionDriver(pin,pm,pout) {
   pmesh = pm;
   std::string integrator_name = pin->GetOrAddString("time", "integrator", "rk2");
+  int nstages = 2; // default rk2
```
yes

jdolence commented on src/parthenon_manager.hpp:
```diff
@@ -48,7 +48,9 @@ class ParthenonManager {
     ArgParse arg;
     clock_t tstart_;
+#ifdef OPENMP_PARALLEL
     double omp_start_time_;
```
Added an issue. #102

jdolence commented:
@nmsriram @AndrewGaspar @pgrete I think I've addressed your comments sufficiently.  Please let me know if there's anything else you'd like to see done before I pull the trigger.  I feel like other things are waiting on this merge now, so I'll give it 24 hours or wait until you approve, whichever comes first, and then merge.

Thanks again for the useful comments!

jdolence commented:
Thanks everybody!  pulling the trigger...
